### PR TITLE
Adding SageMakerEmbeddingEndPoint

### PR DESCRIPTION
**Pull request by iron8kid**
# Description

- Adding SageMakerEmbeddingEndPoint to embeddings
- Making some corrections to SageMakerLLMEndPoint
- Adding a notebook showcasing the SageMakerEmbeddingEndPoint

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -47,6 +47,8 @@ version = "0.9.35"
 [tool.poetry.dependencies]
 SQLAlchemy = {extras = ["asyncio"], version = ">=1.4.49"}
 beautifulsoup4 = {optional = true, version = "^4.12.2"}
+botocore = ">=1.33.13"
```
Comment (hatianzhang) [inline pyproject.toml:50]: might not want to add dep here for those 2

File: llama_index/utilities/aws_utils.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+from typing import Optional
+
+import boto3
```
[the remark this answers is not in the document]
Comment (iron8kid) [reply llama_index/utilities/aws_utils.py:3]: botocore is needed also to declare the get_aws_service_client output :/

File: pyproject.toml (hunk 1 above)
Comment (iron8kid) [reply pyproject.toml:50]: As they are not lazily imported, they should be added; otherwise, tests will generate errors.

File: pyproject.toml [hunk 3]
```diff
@@ -111,6 +113,7 @@ query_tools = [
 [tool.poetry.group.dev.dependencies]
 black = {extras = ["jupyter"], version = "<=23.9.1,>=23.7.0"}
 boto3 = "1.33.6"  # needed for tests
```
Comment (hatianzhang) [inline pyproject.toml:115]: boto3 already import here

File: pyproject.toml (hunk 1 above)
Comment (hatianzhang) [reply pyproject.toml:50]: oh you mean unit tests?

File: llama_index/embeddings/sagemaker_embedding_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    BaseIOHandler,
+    IOHandler,
+)
+from llama_index.types import PydanticProgramMode
+from llama_index.utilities.aws_utils import get_aws_service_client
+
+DEFAULT_IO_HANDLER = IOHandler()
+
+
+class SageMakerEmbeddingEndPoint(BaseEmbedding):
```
Comment (logan-markewich) [inline llama_index/embeddings/sagemaker_embedding_endpoint.py:17]: tbh, I don't love these class names. I would have preferred `SagemakerLLM` and `SagemakerEmbedding` for example. I might push a change to fix this 😆

Comment (iron8kid) [reply llama_index/embeddings/sagemaker_embedding_endpoint.py:17]: feel free to do so :)

File: llama_index/utilities/aws_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (logan-markewich) [reply llama_index/utilities/aws_utils.py]: This can still be optional (and should be), made a change :)

File: pyproject.toml (hunk 1 above)
Comment (logan-markewich) [reply pyproject.toml]: Both can be in dev dependencies, no need to add to core deps

File: llama_index/embeddings/sagemaker_embedding_endpoint.py (hunk 4 above)
Comment (logan-markewich) [reply llama_index/embeddings/sagemaker_embedding_endpoint.py]: done (and kept the old LLM name for backwards compat)

**Comment by iron8kid** [comment]
@logan-markewich  Thank you  :)

**Review by logan-markewich — APPROVED** [review]

**Pull request merged**